### ISSUE TITLE
Appending dependency implementations via handlers.

### ISSUE THREAD

**Issue by godrose** [simpleinjector/SimpleInjector#203]
I have a plugin/module based desktop application where the modules are discovered/added dynamically.
After the bootstrap phase they are injected into the ``ShellViewModel``.
I am wondering how can I register the modules in the container such that when the collection is requested it gets resolved correctly.
something like ``container.AppendHandler(typeof(IModule), c => instance);`` This code gets called inside the module of course.

**Comment by TheBigRic** [comment]
> I have a plugin/module based desktop application where the modules are discovered/added dynamically.
They will discovered during the lifetime of the application or only in the initialization phase of the application?

**Comment by godrose** [comment]
Right now only during init phase.
But that's an interesting point you brought up.

**Comment by TheBigRic** [comment]
> I am wondering how can I register the modules in the container such that when the collection is requested it gets resolved correctly.

If it is only during the init phase at this moment, what is the exact problem you have? Sorry, but I don't think I completely understand your question

**Comment by godrose** [comment]
I'll rephrase ))
The application has a navigation-bar-like component which displays the list of pages/tabs. 
Each one of the pages corresponds to a UI Module which is defined in a separate assembly.
These modules are discovered and registered into the IoC container during the bootstrap phase of the application.
They are registered into the Container so that when the ``ShellViewModel`` is resolved the **collection** of these modules is injected into it and then it gets presented in the nav-bar.
Hope that clarifies what I'm trying to achieve.
P.S. I can solve this with MEF but I want to have cleaner solution via the IoC container.

**Comment by godrose** [comment]
The code for registration for each module goes like this:
``iocContainer.RegisterHandler(typeof(IUiModule), null, c => this);``
and for injection:
``public ShellViewModel (             
            IEnumerable<IUiModule> modules)
        {           
            _rawModules = modules;         
        }``
Currently it works only thanks to the implementation of the IoC container I'm using.
I'd like to use **SimpleInjector** but cannot find equivalent functionality.

**Comment by qujck** [comment]
There is [SimpleInjector.Packaging](https://github.com/simpleinjector/SimpleInjector/tree/master/src/SimpleInjector.Packaging) but read [this SO answer](http://stackoverflow.com/a/25109306) to see there are a number of simple ways to roll your own.

**Comment by TheBigRic** [comment]
To inject a IEnumerable<T> into some component you need to register it as a 'Collection', using the `RegisterCollection` method as described here: [Resolve arrays and lists](http://simpleinjector.readthedocs.org/en/latest/howto.html#resolve-arrays-and-lists)

You must understand that [Simple Injector differentiates between 'normal' registrations and the registration of collections](http://simpleinjector.readthedocs.org/en/latest/decisions.html?highlight=registercollection#separate-collections)

There is also a 'Plugin' extension which is described here: [Registering plugins dynamically](http://simpleinjector.readthedocs.org/en/latest/advanced.html#registering-plugins-dynamically)

**Comment by godrose** [comment]
@TheBigRic I can't use the ``RegisterCollection`` because i don't know the exact types of the UI modules. So my approach is to separate the registrations instead of concentrating them in one place like in ``RegisterCollection`` case.
The plugin case looks more suitable but again I won't be able to distribute the registrations to the modules themselves.
I guess the second link you provided highlights the difference in the approaches.

**Comment by godrose** [comment]
@qujck Thanks for the link. Again this is something I already do for existing plugins in very similar way. 
The root cause of the issue is the need to **inject** the UI Modules into the ``ShellViewModel``.
So the question is how to bridge over this gap.

**Comment by qujck** [comment]
I think I've misunderstood what you are trying to do and have no idea what you are trying to do all at the same time!

**Comment by godrose** [comment]
@TheBigRic  @qujck  Thanks for the help guys. I think that MEF is more appropriate for what I'm doing.

**Comment by dotnetjunkie** [comment]
There is a section in the docs particularly about registering plugins: [here](https://simpleinjector.readthedocs.org/en/latest/advanced.html#plugins).

Does that section answer your question?

Note that you typically want to load plugins at startup in a desktop app; not dynamically on the fly. Especially not being able to hot-swap plugin assemblies. That adds a lot of complexity and should only be done when absolutely required. Otherwise, just let the user restart the application.

**Comment by TheBigRic** [comment]
> I think that MEF is more appropriate for what I'm doing.

Why?

>  I can't use the  RegisterCollection  because i don't know the exact types of the UI modules

I don't understand, if you want this:
``` c#
 public ShellViewModel (IEnumerable<IUiModule> modules)
 { 
    _rawModules = modules; 
 } 
```
Than you would typically register this as:
``` c#
   container.RegisterCollection(typeof(IUITModule), listOfAssemblies);
```
The listOfAssemblies can just be the list of your dynamically discovered assemblies.

So can you elaborate?

**Comment by godrose** [comment]
@BigRic Right now I have an architecture which is agnostic to which IoC container I use. I want to keep it that way. The last suggestion is definitely worth a shot but it requires another abstraction in the architecture. I will test it.
Thanks a lot!

**Comment by qujck** [comment]
You also have [`AppendToCollection`](http://simpleinjector.readthedocs.org/en/latest/advanced.html#mixing-collections-of-open-generic-and-non-generic-components) available to you in the `SimpleInjector.Advanced` namespace.

**Comment by godrose** [comment]
Thanks everyone. After some adjustments I've managed to use SimpleInjector in my project.
I love when people just give a new perspective thus making things much clearer.
In case you're interested here's the link to what I've been doing:
https://github.com/LogoFX

**Comment by dotnetjunkie** [comment]
Please be aware that [this code](https://github.com/LogoFX/LogoFX/blob/master/LogoFX.Client.Bootstrapping.Adapters.SimpleInjector/SimpleInjectorAdapter.cs) is effectively an implementation of the [Conforming Container anti-pattern](http://blog.ploeh.dk/2014/05/19/conforming-container/). Please try to understand what this is and why it is an anti-pattern. If you are building a reusable library, please take a look at [this article](http://blog.ploeh.dk/2014/05/19/di-friendly-library). The article describes how to design a library without the need of a Conforming Container.

**Comment by godrose** [comment]
@dotnetjunkie Thanks for the links.
After reading this I must say I do not quite agree with all points mentioned there.
Calls to the Conforming Container are likely to be sprinkled liberally over an entire code base - No, if you use ctor injection and perform the registrations inside the Composition Root.
It pushes novice users towards the Service Locator anti-pattern. Most people encountering Dependency Injection for the first time mistake it for the Service Locator anti-pattern, despite the entirely opposite natures of these two approaches to loose coupling - That's true for any IoC container.
It attempts to relieve symptoms of bad design, instead of addressing the underlying problem. Too many 'loosely coupled' designs attempt to rely on the Service Locator anti-pattern, which, by default, introduces a dependency to a concrete Service Locator throughout a code base. However, exclusively using the Constructor Injection and Composition Root design patterns eliminate the problem altogether, resulting in a simpler design with fewer moving parts - Again, one may abuse any IoC Container and use is as Service Locator. Having a common interface should not influence this decision.
It pulls in the direction of the lowest common denominator - Yes, it's a drawback, but you can overcome this in the correspondent adapter. Unity does not possess all the features SimpleInjector has to offer, so all advanced registration activities should be performed in the Composition Root and thus encapsulated.
It stifles innovation, because new, creative, but radical ideas may not fit into the narrow view of the world a Conforming Container defines - bunch of abstract stuff, really. You can **always** implement this in the actual IoC container.
It makes it more difficult to avoid using a DI Container. A DI Container can be useful in certain scenarios, but often, hand-coded composition is better than using a DI Container. However, if a library or framework depends on a Conforming Container, it may be difficult to harvest the benefits of hand-coded composition - what? How does it make it more difficult. Having this interface does not mean I **have** to use it, it's up to me and how i design the bootstrapper.
It may introduce versioning hell. Imagine that you need to use a library that depends on Confainer 1.3.7 in an application that also uses a framework that depends on Confainer 2.1.7. Since a Conforming Container is intended as an infrastructure component, this is likely to happen, and to cause much grief - true for any library/package.
A Conforming Container is often a product of Speculative Generality, instead of a product of need. As such, the API is likely to be poorly suited to address real-world scenarios, be difficult to extent, and may exhibit churn in the form of frequent breaking changes - yes, it may. Will it eventually happen? It may and it may not.
If Adapters are supplied by contributors (often the DI Container maintainers themselves), the Adapters may have varying quality levels, and may not support the latest version of the Conforming Container - Correct, that's a major issue right now.

To sum up, this does not look very convincing, and the benefits so far outweigh the disadvantages.

Thanks again for the useful links.

**Comment by TheBigRic** [comment]
The Conforming Container anti-patterns consists of two parts. 
 1. a Service Locator, which is already known to be an [anti-pattern]( http://blog.ploeh.dk/2010/02/03/ServiceLocatorisanAnti-Pattern/).
 2. An registration API, which we feel actually is an anti-pattern as well.

Not all arguments of Mark Seemann are evenly strong, but be careful to dismiss the whole argument just by weighing the pros and the cons.

Some feedback on your statements:

> It pulls in the direction of the lowest common denominator - Yes, it's a drawback, but you can overcome this in the correspondent adapter.
 
You may want to take a look at this [discussion](https://github.com/simpleinjector/SimpleInjector/issues/41#issuecomment-111438146) where @dotnetjunkie sums up a list of problems that is currently present in the `Conforming container` is Asp.Net vNext. As you can see, there is no such thing as 'but you can overcome this in the correspondent adapter', maybe the simple things but certainly not all.
 
One of the big points here is expectation of the user. I will try to make this clear by this example:
 
Let's say I would use your library and thereby use the Simple Injector adapter. Because I used Simple Injector many times before, I have some expectations:
 
1. If I make a mistake, e.g. I forgot to register some component, my application will break at startup, because I call `Verify()`, which will detect this for me. Your adapter does not do that.
2. Let's say we accept the fact that `Verify()` isn't called, because calling `GetInstance()` will also throw an exception (with a very clear message what it is exactly that I have done wrong) if I made a mistake. Your adapter does not do that, instead it returns null, resulting in a `NullReferenceExceptioin` somewhere deep in the stack.
3. Imagine a I have a very big object graph build up from all singletons. Then I add a component to the object graph (just by adding it to some constructor of some component). But this component is also used in some other part of the application in an object graph which has a `Scoped lifestyle`. Simple Injector will normally (by calling verify or retrieving an instance) throw an activation exception, explaining me very clearly what I have done wrong ([Lifestyle Mismatch]( http://simpleinjector.readthedocs.org/en/latest/LifestyleMismatches.html)). Your adapter however will break the application by throwing a `NullReferenceException` somewhere deep in the stack because the exception is catched and `null` is returned.
4. Because of the excellent support for generics in SimpleInjector, I design my application around open generic interfaces. Registering this with Simple Injector is a one liner. Many other DI containers do this differently or lack support for this. Your abstraction and thus adapter is missing this feature completely.
5. Using decorators, implementing cross cutting concerns is simple. Decorators are completely unsupported by your (and other) conforming container(s). 
 
All points I mentioned, and I could go on, are the reason I use Simple Injector. Just point 2, where your adapter fails silently instead of [failing fast](https://en.wikipedia.org/wiki/Fail-fast) would lead to the decision not to use whatever (conforming) container is presented to me. And I'm guessing this is true for 80% of the Simple Injector users. All users from whatever container depend on some feature that isn't the part of the 'lowest common denominator'. So this probably holds also for the users of AutoFac and Ninject.

> It stifles innovation, because new, creative, but radical ideas may not fit into the narrow view of the world a Conforming Container defines bunch of abstract stuff, really. You can always implement this in the actual IoC container.

Here again #41 is a good example of this. ASP.NET forces us into a model that excludes Simple Injector. In a sense SI is 'radical' because it differentiates from 'the narrow view of the world' of ASP.NET's conforming container. On the other hand, if you would design a conforming container with SI in mind, you will exclude some current or future containers, and probably [Pure DI](http://blog.ploeh.dk/2014/06/10/pure-di/) practitioners. 

 > It makes it more difficult to avoid using a DI Container

This statement from Seemann is actually true and we've seen this with frameworks like NServiceBus and NancyFx for instance that apply Conforming Containers. Applying Pure DI has proven to be hard (NancyFx) to impossible (NServiceBus). We've had numerous discussions with those teams, which made both us and those teams come to that conclusion. NancyFx team even explicitly stated that Nancy wasn't designed for Pure DI, which basically means they exclude that group of users and type of development completely. NServiceBus is fortunately planning to step away from the conforming container pattern because of the downsides it has.

 > It may introduce versioning hell => true for any library/package.

Yes, but with the conforming container you quadruple the risk, because you are both dealing with the conforming container, the adapters and the containers. To take Simple Injector for instance, in v3 we made quite some changes that broke almost all conforming container adapters out there. Multiply this by the number of containers, their versions, adapters and their versions and you'll start to see the problem.

The thing is, if you design your application correctly, which seems you're already doing, the need for a conforming container and all adapters is pretty limited. The only place in the application where the code is dependent on a certain DI container is the composition root. And if the library is `DI friendly` as described in the linked blogpost from Mark Seeman, configuring the container for your library should be a walk in the park.
 
So which problem does a conforming container really solve?

**Comment by godrose** [comment]
First of all thanks for investing your time to explain your point clearly. I appreciate that a lot.

You're right I do use the ``Inversion-Of-Control`` container only in the ``Composition Root`` part of the app. 
I'm mostly involved in the Client-side development in Web and .NET/UWP.
I have developed an approach to designing and implementing apps which involves lots of infrastructure code that's encapsulated in the correspondent packages. This involves the ``Bootstrapping`` aspect which performs some common UI-related tasks while remaining agnostic to the actual IoC container.
So instead of having several types of bootstrappers per each container (as I used to have in the past) I have one and I can substitute the adapters.
This saves a lot of time and keeps the bootstrapping code stable and independent of the concrete container API's changes.
If there's some API-specific stuff to be registered for the given container - ``Decorators`` in ``SimpleInjector`` etc. you can always instantiate the concrete container, configure it and then just pass the configured ``SimpleInjector Container`` to the correspondent ``Adapter``. I'm not making you configure the container via its adapter - that's pointless, of course.

Another point in favor of the ``Adapters`` is the fact that they also provide an adapter to the ``Caliburn.Micro`` required IoC functionality. Hence the ``Adapters`` effectively do two things: they provide container-specific implementation for the common ``IIocContainer`` interface and for the ``Caliburn.Micro`` related ``IBootstrapperAdapter``. I know, it's not the cleanest solution possible, but I felt that having something like ``Bootstrapping.Adapters.Caliburn.Micro.SimpleInjector`` package is just an overkill...

I think that by working more with ``SimpleInjector`` I will be able to discover the missing parts in my approach and fill them up appropriately.

**Comment by godrose** [comment]
I would love to see an enterprise-scale app which fully employs the features of the ``SimpleInjector``. Then I'll be in better position to full understand your concerns.

**Comment by dotnetjunkie** [comment]
Those are examples that are hard to find online, even with other DI libraries. I've assisted multiple teams the last years with 'enterpricy' (what's the definition of enterprise-scale anyway?) apps, and what I teach them and all those apps work around a few simple message based concepts as the onces I describe on my blog. You can built the biggest systems with those patterns and IMO those patterns are a requirement for maintainable software at large scale. And you know what? Simple Injector is especially suited to wite up applications that are designed around these patterns.
